### PR TITLE
Reference table Phase-1

### PR DESCRIPTION
**Pull request by onderkalaci**
@marcocitus This PR implements partially #969. 

With this commit, we implemented some basic features of reference tables.

To start with, a reference table is
  * a distributed table whithout a distribution column defined on it
  * the distributed table is single sharded
  * and the shard is replicated to all nodes

Reference tables follows the same code-path with a single sharded
tables. Thus, broadcast JOINs are applicable to reference tables.
But, since the table is replicated to all nodes, table fetching is
not required any more.

Reference tables support the uniqueness constraints for any column.

Reference tables can be used in INSERT INTO .. SELECT queries with
the following rules:
  * If a reference table is in the SELECT part of the query, it is
    safe join with another reference table and/or hash partitioned
    tables.
  * If a reference table is in the INSERT part of the query, all
    other participating tables should be reference tables.

Reference tables follow the regular co-location structure. Since
all reference tables are single sharded and replicated to all nodes,
they are always co-located with each other.

Queries involving only reference tables always follows router planner
and executor.

Reference tables can have composite typed columns and there is no need
to create/define the necessary support functions.

All modification queries, master_* UDFs, EXPLAIN, DDLs, TRUNCATE,
sequences, transactions, COPY, schema support works on reference
tables as expected. Plus, all the pre-requisites associated with
distribution columns are dismissed.

@marcocitus I was planning to have more granular commits, but, ended up with a huge commit. Does that work for you :) 

I'll be testing the changes some more, but I don't think that is anything blocking you to start the review.

### CODE REVIEW

File: src/backend/distributed/commands/create_distributed_table.c [hunk 1]
```diff
@@ -153,22 +171,40 @@ Datum
...
 	int shardCount = 1;
-	AttrNumber firstColumnAttrNumber = 1;
+	int replicationFactor = list_length(workerNodeList);
+	Oid distributionColumnType = InvalidOid;
+	char *distributionColumnName = NULL;
 
-	char *firstColumnName = get_attname(relationId, firstColumnAttrNumber);
-	if (firstColumnName == NULL)
+	/* check for existing colocations */
+	colocationId = ColocationId(shardCount, replicationFactor, distributionColumnType);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: > Will we continuously update the replication factor in pg_dist_colocation for reference tables? What if we have a special case, e.g. replication factor 0, or a reserved colocation group ID?

That's a good point. Since we only need to update when a node is added, I'm planning to handle  that #971. Does that sound OK to you?

File: src/include/distributed/pg_dist_partition.h [hunk 2]
```diff
@@ -52,6 +52,7 @@ typedef FormData_pg_dist_partition *Form_pg_dist_partition;
 #define DISTRIBUTE_BY_APPEND 'a'
 #define DISTRIBUTE_BY_HASH 'h'
 #define DISTRIBUTE_BY_RANGE 'r'
+#define DISTRIBUTE_BY_ALL 'l'
```
Comment (anarazel) [inline src/include/distributed/pg_dist_partition.h]: Alternatively we could start using the enum directly, that'd certainly be easier to read from SQL. Albeit a bit more annoying from C.

File: src/backend/distributed/utils/metadata_cache.c [hunk 3]
```diff
@@ -1460,8 +1487,11 @@ ResetDistTableCacheEntry(DistTableCacheEntry *cacheEntry)
 		cacheEntry->hasUninitializedShardInterval = false;
 		cacheEntry->hasUniformHashDistribution = false;
 
-		pfree(cacheEntry->shardIntervalCompareFunction);
-		cacheEntry->shardIntervalCompareFunction = NULL;
+		if (cacheEntry->partitionMethod != DISTRIBUTE_BY_ALL)
```
Comment (anarazel) [inline src/backend/distributed/utils/metadata_cache.c]: I'd rather make the if about shardIntervalCompareFunction being non-null.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 4]
```diff
@@ -755,6 +786,64 @@ ErrorIfMultiTaskRouterSelectQueryUnsupported(Query *query)
...
+	}
+}
+
+
+/*
+ * TODO: update comment
+ */
+bool
+AllParticipatingTablesAreReferenceTables(Query *query)
+{
```
Comment (anarazel) [inline src/backend/distributed/planner/multi_router_planner.c]: I'm not saying we *have* to do that, but we could very well integrate this into the pass already done via multi_relation_restriction_hook. RelationRestrictionContext already has hasDistributedRelation, we could easily (and cheaply!) add this knowledge, without additional passes over the parsetree.

File: src/backend/distributed/Makefile [hunk 5]
```diff
@@ -94,6 +94,8 @@ $(EXTENSION)--6.0-17.sql: $(EXTENSION)--6.0-16.sql $(EXTENSION)--6.0-16--6.0-17.
 	cat $^ > $@
 $(EXTENSION)--6.0-18.sql: $(EXTENSION)--6.0-17.sql $(EXTENSION)--6.0-17--6.0-18.sql
 	cat $^ > $@
+$(EXTENSION)--6.0-19.sql: $(EXTENSION)--6.0-18.sql $(EXTENSION)--6.0-18--6.0-19.sql
```
Comment (onderkalaci) [inline src/backend/distributed/Makefile]: > Should start with 6.1.

I realized that, but, left it since I'll most probably rebase and have that :)

File: src/backend/distributed/utils/metadata_cache.c (hunk 3 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/metadata_cache.c]: That makes sense, I'll update.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 4 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c]: > I'm not saying we have to do that, but we could very well integrate this into the pass already done via multi_relation_restriction_hook. RelationRestrictionContext already has hasDistributedRelation, we could easily (and cheaply!) add this knowledge, without additional passes over the parsetree.

Thanks for noting that. I agree that it makes sense to have that information there. That'll help to remove some code from the PR as well.

File: src/backend/distributed/commands/create_distributed_table.c [hunk 6]
```diff
@@ -153,22 +171,40 @@ Datum
...
-						errmsg("reference table candidate %s needs to have at"
-							   "least one column", relationName)));
+		colocationId = CreateColocationGroup(shardCount, replicationFactor,
+											 distributionColumnType);
 	}
 
-	CreateHashDistributedTable(relationId, firstColumnName, shardCount,
-							   ShardReplicationFactor);
+	/* first, convert the relation into distributed relation */
+	ConvertToDistributedTable(relationId, distributionColumnName,
```
Comment (marcocitus) [inline src/backend/distributed/commands/create_distributed_table.c]: Well, the difference is that we could do '2' (statement-based using 2PC) using MX, but not 'c' (statement-based using coordinator).

Comment (marcocitus) [inline src/backend/distributed/commands/create_distributed_table.c]: Which is relevant, since we use the repmodel to know which tables we can sync in MX.

**Comment by marcocitus** [comment]
> Alternatively we could start using the enum directly, that'd certainly be easier to read from SQL. Albeit a bit more annoying from C.

One thing we observed is that adding an option to the enum would involve dropping the enum, which would then cascade to create_distributed_table and potentially user-defined wrappers.

File: src/include/distributed/pg_dist_partition.h (hunk 2 above)
Comment (marcocitus) [inline src/include/distributed/pg_dist_partition.h]: We'd have to drop the existing enum, which would have to cascade to `create_distributed_table`, which might cascade to user-defined wrappers.

**Comment by anarazel** [comment]
On 2016-12-13 05:52:27 -0800, Marco Slot wrote:
> > Alternatively we could start using the enum directly, that'd certainly be easier to read from SQL. Albeit a bit more annoying from C.
> 
> One thing we observed is that adding an option to the enum would involve dropping the enum, which would then cascade to create_distributed_table and potentially user-defined wrappers.

You can add a value (ALTER TYPE .. ADD VALUE ...) , but admittedly its
restricted due to some transactional requirements.

File: src/backend/distributed/commands/create_distributed_table.c (hunk 6 above)
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: After our chat in person, I'm more convinced that it'd be useful to have a different replication model.
I'm picking `t` which could stand for `transactional` or `two-phase commit` :)

File: src/backend/distributed/planner/multi_join_order.c [hunk 7]
```diff
@@ -1276,7 +1292,8 @@ SinglePartitionJoin(JoinOrderNode *currentJoinNode, TableEntry *candidateTable,
 	}
 
 	/* evaluate re-partitioning the current table only if the rule didn't apply above */
-	if (nextJoinNode == NULL && candidatePartitionMethod != DISTRIBUTE_BY_HASH)
+	if (nextJoinNode == NULL && candidatePartitionMethod != DISTRIBUTE_BY_HASH &&
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_join_order.c]: @marcocitus This is one of the places where I'm not sure whether the change is correct.
In general, I don't understand why we have `candidatePartitionMethod != DISTRIBUTE_BY_HASH` check here. Could you pay some extra attention to this change?

File: src/backend/distributed/utils/distribution_column.c [hunk 8]
```diff
@@ -129,6 +132,12 @@ BuildDistributionKeyFromColumnName(Relation distributedRelation, char *columnNam
 	Var *distributionColumn = NULL;
 	char *tableName = RelationGetRelationName(distributedRelation);
 
+	/* short circuit for reference tables */
```
Comment (onderkalaci) [inline src/backend/distributed/utils/distribution_column.c]: Just a note for the reviewer. I see that this is a hacky solution. However, my other attempts to refactor the function lead to many other cascading changes.The main reason is that once the function is called, `distributedRelation` is not actually distributed. So, it is not possible to learn the distribution method. Are you OK with that?

File: src/backend/distributed/planner/multi_join_order.c (hunk 7 above)
Comment (marcocitus) [inline src/backend/distributed/planner/multi_join_order.c]: We only do range-repartitioining for single-repartition joins and hash-repartitioning for dual-repartition joins. So when joining two hash-distributed tables, we always re-partition both even if one is partitioned on the join column. This is something we should consider fixing.

Comment (onderkalaci) [inline src/backend/distributed/planner/multi_join_order.c]: Then, I *think* the change I did makes sense here as well.

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 9]
```diff
@@ -2580,6 +2582,14 @@ PruneShardList(Oid relationId, Index tableId, List *whereClauseList,
 	Var *partitionColumn = PartitionColumn(relationId, tableId);
 	char partitionMethod = PartitionMethod(relationId);
 
+	/* short circuit for reference tables */
+	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		Assert(list_length(shardIntervalList) == 1);
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_physical_planner.c]: I feel like we tend to overuse Assert. Incorrect metadata shouldn't really crash the server, but rather cause errors (e.g. someone may have a bug in a metadata-changing script). We should error out somewhere if a reference table has >1 shard, though probably closer to where we load shard list.

File: src/backend/distributed/utils/distribution_column.c [hunk 10]
```diff
@@ -129,6 +132,12 @@ BuildDistributionKeyFromColumnName(Relation distributedRelation, char *columnNam
...
+	if (columnName == NULL)
+	{
+		return distributionColumn;
```
Comment (marcocitus) [inline src/backend/distributed/utils/distribution_column.c]: Maybe change to `return NULL` to make it clearer.

File: src/backend/distributed/utils/distribution_column.c (hunk 8 above)
Comment (marcocitus) [inline src/backend/distributed/utils/distribution_column.c]: I guess the alternative is to not call it for references tables, e.g.

```
    /* only build a distribution column for distributed tables, not reference tables */
    if (distributionMethod != DISTRIBUTE_BY_NONE)
    {
        distributionColumn = BuildDistributionKeyFromColumnName(relation,
                                                                distributionColumnName);
    }
```
in `create_distributed_table.c`, which would be clearer to the caller.

Comment (onderkalaci) [inline src/backend/distributed/utils/distribution_column.c]: > I guess the alternative is to not call it for references tables

In general*, I followed the idea that all functions can be called with reference tables. That's why I proposed the change as this here. But, I'm cool with changing if you 

*I had to have an exception for `AddShardIntervalRestrictionToSelect()` which is not possible to call on a reference table and was hard to refactor.

File: src/backend/distributed/planner/multi_physical_planner.c (hunk 9 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c]: > I feel like we tend to overuse Assert.

Hmm, that makes sense. I mostly try to add many `Asserts` during the development and remove on the PR. 

For this PR, I'll consider removing some of the `Asserts` I added.

**Comment by onderkalaci** [comment]
@marcocitus One more note on the PR. Also, it'd be great to hear your comment on this.

As far as I can see, we currently do not support broadcast join / small tables for subquery push down. That's why whenever a reference table involved in a query, `ErrorIfCannotPushdownSubquery()` errors out and cannot proceed to logical planner / optimizer. Also, even if it **somehow** could pass through the error checks, `ErrorIfUnsupportedShardDistribution()` function would error out. 

The issue is that there are few functions (i.e., `SubquerySqlTaskList()`) in the further code-path that expects partition column to exists. I think that it is not in the scope of this PR to enable reference tables for subquery push-down. Do you agree with that?

**Comment by marcocitus** [comment]
Subquery pushdown might not give correct results when a reference table is involved, e.g. when performing [reference table] left join [distributed table], so I think that's why it's disallowed.

**Comment by onderkalaci** [comment]
@marcocitus I finished my tests, rebased to master (fixed bunch of merge conflicts with `colocate_with` PR). It's now ready for your review.

Thanks!

**Comment by marcocitus** [comment]
The 'n' actually looks a lot like the 'h' :-/

```
# SELECT logicalrelid, partmethod, column_to_column_name(logicalrelid, partkey) FROM pg_dist_partition;
 logicalrelid | partmethod | column_to_column_name 
--------------+------------+-----------------------
 ref          | n          | 
 test         | h          | x
```

**Comment by marcocitus** [comment]
The error message during shard creation is a bit confusing:

```
postgres=# SET citus.shard_replication_factor TO 1;
SET
Time: 0.779 ms
postgres=# SELECT create_reference_table('ref2');
WARNING:  connection error: localhost:9700
DETAIL:  could not connect to server: Connection refused
        Is the server running on host "localhost" (127.0.0.1) and accepting
        TCP/IP connections on port 9700?

WARNING:  could not create shard on "localhost:9700"
ERROR:  could only create 1 of 2 of required shard replicas
```

File: src/test/regress/expected/multi_explain.out [hunk 11]
```diff
@@ -497,19 +497,16 @@ EXPLAIN (COSTS FALSE)
...
   Executor: Task-Tracker
   Task Count: 1
   Tasks Shown: None, not supported for re-partition queries
   ->  MapMergeJob
-        Map Task Count: 1
+        Map Task Count: 8
         Merge Task Count: 1
-        ->  MapMergeJob
-              Map Task Count: 8
-              Merge Task Count: 1
```
Comment (marcocitus) [inline src/test/regress/expected/multi_explain.out]: We're losing test coverage on explaining nested jobs here, could we get that back somehow?

Comment (onderkalaci) [inline src/test/regress/expected/multi_explain.out]: To be more specific, because we've replaced `supplier` to a reference table, it is always planned to be a broadcast join. 
What about  I create another table for this specific test say `supplier_single_shard_hash` and use it? Does that work for you?

Comment (marcocitus) [inline src/test/regress/expected/multi_explain.out]: sure

**Comment by onderkalaci** [comment]
>Testing with an empty cluster is a little weird right now, should maybe error out in create_reference_table.

That makes sense.

I added an error message for creating reference when there are no worker nodes.

> The error message during shard creation is a bit confusing:

Isn't this the same for all partition types?

**Comment by marcocitus** [comment]
> Isn't this the same for all partition types?

Yes, but a user creating a reference tale might not know why "1 of 2 shard replicas" is relevant or what do about it.

**Comment by onderkalaci** [comment]
> Yes, but a user creating a reference tale might not know why "1 of 2 shard replicas" is relevant or what do about it.

You're right, it'd be nice to see error out such that `cannot create reference table - not all workers are running` (or so).

But, the error is coming from `CreateReferenceTable()->CreateReferenceTableShard()->CreateShardPlacements()`

So, I guess we may need to refactor a lot of code. Do you want me to go on?

File: src/backend/distributed/master/master_create_shards.c [hunk 12]
```diff
@@ -322,6 +322,93 @@ CreateColocatedShards(Oid targetRelationId, Oid sourceRelationId)
...
+						errmsg("table \"%s\" has already had shards created for it",
+							   tableName)));
+	}
+
+	/* load and sort the worker node list for deterministic placement */
+	workerNodeList = WorkerNodeList();
+	workerNodeList = SortList(workerNodeList, CompareWorkerNodes);
+
+	/* make sure we don't process cancel signals until all shards are created */
+	HOLD_INTERRUPTS();
```
Comment (onderkalaci) [inline src/backend/distributed/master/master_create_shards.c]: But, as far as I can see, we don't use it yet for shard creation?

Comment (marcocitus) [inline src/backend/distributed/master/master_create_shards.c]: We do. Everything uses the connection API.

**Comment by marcocitus** [comment]
> So, I guess we may need to refactor a lot of code. Do you want me to go on?

Hm, not very much.

File: src/backend/distributed/master/master_create_shards.c (hunk 12 above)
Comment (onderkalaci) [inline src/backend/distributed/master/master_create_shards.c]: Ok. Also, `CreateShardsWithRoundRobinPolicy` uses it as well. Should I remove that too in this PR?

**Comment by onderkalaci** [comment]
> Hm, not very much.

Is it you don't want it much or the refactor doesn't look to require much change :bowtie:

**Comment by marcocitus** [comment]
> Is it you don't want it much or the refactor doesn't look to require much change :bowtie:

Don't want it much :)

File: src/backend/distributed/master/master_create_shards.c (hunk 12 above)
Comment (marcocitus) [inline src/backend/distributed/master/master_create_shards.c]: No, seems unrelated, we can clean that up seperately.

File: src/backend/distributed/commands/create_distributed_table.c [hunk 13]
```diff
@@ -744,6 +803,10 @@ LookupDistributionMethod(Oid distributionMethodOid)
 	{
 		distributionMethod = DISTRIBUTE_BY_RANGE;
 	}
+	else if (strncmp(enumLabel, "all", NAMEDATALEN) == 0)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: That makes sense, we don't need this

File: src/backend/distributed/commands/create_distributed_table.c [hunk 14]
```diff
@@ -344,10 +388,25 @@ static void
...
+	char *relationName = NULL;
+	List *indexOidList = NULL;
 	ListCell *indexOidCell = NULL;
 
+	/*
+	 * Citus supports any kind of uniqueness constraints for reference tables
+	 * given that they only consist of a single shard and we can simply rely on
+	 * Postgres.
+	 * TODO: Here we should be erroring out if there exists any foreign keys
+	 * from/to a reference table.
```
Comment (marcocitus) [inline src/backend/distributed/commands/create_distributed_table.c]: Will this be in a separate PR?

**Comment by marcocitus** [comment]
Reference tables reveal some distributed deadlocks in the router executor.

In session 1:
```
CREATE TABLE test (x int, y int);
SELECT create_distributed_table('test', 'x');

CREATE TABLE ref (y int primary key, name text);
SELECT create_reference_table('ref');

INSERT INTO ref VALUES (3,4);
INSERT INTO test VALUES (3,4);

BEGIN;
UPDATE ref SET y = 3;
```

In session 2:
```
BEGIN;
UPDATE test SET y = 5 WHERE x = 3;                                                                                                                 UPDATE ref SET y = 10;
<hangs on shard resource lock held by session 1 on master>
```

In session 1:
```
UPDATE test SET y = 10 WHERE x = 3;
<hangs on row-level lock held by session 2 on worker 1>
```

**Comment by marcocitus** [comment]
I think the PR is ready to 🚢 , though we'll need to do something about the deadlock.
